lustre/doc/lctl.8
.TH lctl 1 "2003 Oct 8" Lustre "configuration utilities"
.SH NAME
lctl \- Low level Lustre filesystem configuration utility
.SH SYNOPSIS
.br
.B lctl
.br
.B lctl --device <devno> <command [args]>
.br
.SH DESCRIPTION
.B lctl
is used to directly control Lustre via an ioctl interface, allowing
various configuration, maintenance, and debugging features to be accessed.
 
.B lctl
can be invoked in interactive mode by issuing lctl command. After that, commands are issued as below. The most common commands in lctl are
.B dl
,
.B dk
,
.B device 
,
.B network 
.I <up/down>
,
.B list_nids
,
.B ping
.I nid
,
.B help
,
.B quit.

To get a complete listing of available commands, type
.B help
at the lctl prompt.  To get basic help on the meaning and syntax of a
command, type 
.B help 
.I command
.  Command completion is activated with the TAB key, and command history is available via the up- and down-arrow keys. 

For non-interactive use, one uses the second invocation, which runs command after connecting to the device. 

.SS Network Configuration
.TP
.BI network " <up/down>|<tcp/elan/myrinet>"
Start or stop LNET, or select a network type for other
.I lctl
LNET commands
.TP
.BI list_nids
Print all Network Identifiers on the local node. LNET must be running.
.TP
.BI which_nid " <nidlist>"
From a list of nids for a remote node, show which interface communication
will take place on.
.TP
.BI ping " <nid> "
Check LNET connectivity via an LNET ping. This will use the fabric
appropriate to the specified NID.
.TP
.BI interface_list 
Print the network interface information for a given 
.B network
type.
.TP
.BI peer_list 
Print the known peers for a given 
.B network
type.
.TP
.BI conn_list 
Print all the connected remote NIDs for a given
.B network
type.
.TP
.BI active_tx 
This command should print active transmits, and it is only used for elan network type.
.TP 
.BI route_list 
Print the complete routing table.
.PP
.SS Device Selection
.TP 
.BI device " <devname> " 
This will select the specified OBD device.  All other commands depend on the device being set. 
.TP 
.BI device_list 
Show all the local Lustre OBDs. AKA 
.B dl
.PP
.SS Device Operations
.TP 
.BI list_param " [-F|-R] <param_search ...>"
List the Lustre or LNet parameter name
.B -F
Add '/', '@' or '=' for dirs, symlinks and writeable files, respectively.
.br
.B -R
Recursively list all parameters under the specified parameter search string. If
.I param_search
is unspecified, all the parameters will be shown.
.br
.B Examples:
.br
.B
# lctl list_param ost.*
.br
  ost.OSS
.br
  ost.num_refs
.br
.B
# lctl list_param -F ost.* debug
.br
  ost.OSS/
.br
  ost.num_refs
.br
  debug=
.br
.B
# lctl list_param -R mdt
.br
  mdt
.br
  mdt.lustre-MDT0000
.br
  mdt.lustre-MDT0000.capa
.br
  mdt.lustre-MDT0000.capa_count
.br
  mdt.lustre-MDT0000.capa_key_timeout
.br
  mdt.lustre-MDT0000.capa_timeout
.br
  mdt.lustre-MDT0000.commit_on_sharing
.br
  mdt.lustre-MDT0000.evict_client
.br
  ...
.TP
.BI get_param " [-n|-N|-F] <parameter ...>"
Get the value of Lustre or LNET parameter.
.br
.B -n
Print only the value and not parameter name.
.br
.B -N
Print only matched parameter names and not the values. (Especially useful when using patterns.)
.br
.B -F
When -N specified, add '/', '@' or '=' for directories, symlinks and writeable files, respectively.
.br
.B Examples:
.br
.B
# lctl get_param ost.*
.br
  ost.OSS
.br
  ost.num_refs
.br
.B
# lctl get_param -n debug timeout
.br
  super warning dlmtrace error emerg ha rpctrace vfstrace config console
.br
  20
.br
.B
# lctl get_param -N ost.* debug
.br
  ost.OSS
.br
  ost.num_refs
.br
  debug
.br
lctl "get_param -NF" is equivalent to "list_param -F".
.TP
.BI set_param " [-n] [-P] [-d] <parameter=value ...>"
Set the value of Lustre or LNET parameter.
.br
.B -n
Disable printing of the key name when printing values.
.br
.B -P
Set the parameter permanently, filesystem-wide.
This parameters are only visible to 2.5.0 and later clients, older clients will not see these parameters.
.br
.B -d
Remove the permanent setting (only with -P option)
.br
.B Examples:
.br
.B
# lctl set_param fail_loc=0 timeout=20
.br
  fail_loc=0
.br
  timeout=20
.br
.B
# lctl set_param -n fail_loc=0 timeout=20
.br
  0
.br
  20
.br
.B
# lctl set_param -P osc.*.max_dirty_mb=32
.br
.TP
.BI conf_param " [-d] <device|fsname>.<parameter>=<value>"
Set a permanent configuration parameter for any device via the MGS.  This
command must be run on the MGS node.
.br
.B -d <device|fsname>.<parameter>
Delete a parameter setting (use the default value at the next restart).  A null value for <value> also deletes the parameter setting.
.br
.B Parameters:
.br
All of the writable parameters under 
.B lctl list_param
(e.g. 
.I lctl list_param -F osc.*.* | grep =
) can be permanently set using
.B lctl conf_param
, but the format is slightly different.  For conf_param, the device is specified first, then the obdtype. (See examples below.)  Wildcards are not supported.
.br
Additionally, failover nodes may be added (or removed), and some system-wide parameters may be set as well (sys.at_max, sys.at_min, sys.at_extra, sys.at_early_margin, sys.at_history, sys.timeout, sys.ldlm_timeout.)  <device> is ignored for system wide parameters.
.br
.B Examples:
.br 
# lctl conf_param testfs.sys.at_max=1200
.br
# lctl conf_param testfs.llite.max_read_ahead_mb=16 
.br
# lctl conf_param testfs-MDT0000.lov.stripesize=2M
.br
# lctl conf_param lustre-OST0001.osc.active=0 
.br
# lctl conf_param testfs-OST0000.osc.max_dirty_mb=29.15 
.br
# lctl conf_param testfs-OST0000.ost.client_cache_seconds=15 
.br
# lctl conf_param testfs-OST0000.failover.node=1.2.3.4@tcp1
.TP 
.BI activate 
Reactivate an import after deactivating, below.  This setting is only effective until the next restart (see 
.B conf_param
).
.TP 
.BI deactivate 
Deactivate an import, in particular meaning do not assign new file stripes
to an OSC.  This command should be used on the OSC in the MDT LOV
corresponding to a failed OST device, to prevent further attempts at
communication with the failed OST.
.TP 
.BI abort_recovery 
Abort the recovery process on a restarting MDT or OST device
.PP
.SS Virtual Block Device Operation
Lustre is able to emulate a virtual block device upon regular file. It is necessary to be used when you are trying to setup a swap space via file.
.TP
.BI blockdev_attach " <file name> <device node>"
Attach the lustre regular file to a block device. If the device node is not existent, lctl will create it \- it is recommended to create it by lctl since the emulator uses a dynamical major number.
.TP
.BI blockdev_detach " <device node>"
Detach the virtual block device.
.TP
.BI blockdev_info " <device node>"
Acquire which lustre file was attached to the device node.
.PP
.SS Changelogs
.TP
.BI changelog_register
Register a new changelog user for a particular device.  Changelog entries
will not be purged beyond any registered users' set point. (See lfs changelog_clear.)
.TP
.BI changelog_deregister " <id>"
Unregister an existing changelog user.  If the user's "clear" record number
is the minimum for the device, changelog records will be purged until the
next minimum.  
.PP
.SS Configuration logs
.TP
.BI replace_nids " <devicename> <nid1>[,nid2,nid3:nid4,nid5:nid6 ...]"
Replace the LNET Network Identifiers for a given device,
as when the server's IP address has changed.
This command must be run on the MGS node.
Only MGS server should be started (command execution returns error
in another cases). To start the MGS service only:
mount -t lustre <MDT partition> -o nosvc <mount point>
Note the replace_nids command skips any invalidated records in the configuration log.
The previous log is backed up with the suffix '.bak'.
Failover nids must be passed after ':' symbol. More then

lnet/ChangeLog
2013-01-22 Oracle, Inc.
	* version 1.8.8
	* Support for networks:
	  socklnd   - any kernel supported by Lustre,
	  qswlnd    - Qsnet kernel modules 5.20 and later,
	  openiblnd - IbGold 1.8.2,
	  o2iblnd   - OFED 1.3, 1.4.1, 1.4.2, 1.5.1 and 1.5.2
	  viblnd    - Voltaire ibhost 3.4.5 and later,
	  ciblnd    - Topspin 3.2.0,
	  iiblnd    - Infiniserv 3.3 + PathBits patch,
	  gmlnd     - GM 2.1.22 and later,
	  mxlnd     - MX 1.2.10 or later,
	  ptllnd    - Portals 3.3 / UNICOS/lc 1.5.x, 2.0.x

-------------------------------------------------------------------------------

2011-12-23 Oracle, Inc.
	* version 1.8.7
	* Support for networks:
	  socklnd   - any kernel supported by Lustre,
	  qswlnd    - Qsnet kernel modules 5.20 and later,
	  openiblnd - IbGold 1.8.2,
	  o2iblnd   - OFED 1.3, 1.4.1, 1.4.2, 1.5.1 and 1.5.2
	  viblnd    - Voltaire ibhost 3.4.5 and later,
	  ciblnd    - Topspin 3.2.0,
	  iiblnd    - Infiniserv 3.3 + PathBits patch,
	  gmlnd     - GM 2.1.22 and later,
	  mxlnd     - MX 1.2.10 or later,
	  ptllnd    - Portals 3.3 / UNICOS/lc 1.5.x, 2.0.x

-------------------------------------------------------------------------------

2011-05-12 Oracle, Inc.
	* version 1.8.6
	* Support for networks:
	  socklnd   - any kernel supported by Lustre,
	  qswlnd    - Qsnet kernel modules 5.20 and later,
	  openiblnd - IbGold 1.8.2,
	  o2iblnd   - OFED 1.3, 1.4.1, 1.4.2, 1.5.1 and 1.5.2
	  viblnd    - Voltaire ibhost 3.4.5 and later,
	  ciblnd    - Topspin 3.2.0,
	  iiblnd    - Infiniserv 3.3 + PathBits patch,
	  gmlnd     - GM 2.1.22 and later,
	  mxlnd     - MX 1.2.10 or later,
	  ptllnd    - Portals 3.3 / UNICOS/lc 1.5.x, 2.0.x

Severity   : normal
Bugzilla   : 24218
Description: fix contention on ksock_tx_t

Severity   : normal
Bugzilla   : 14846
Description: dynamically grow/shrink connd threads pool

Severity   : normal
Bugzilla   : 21776
Description: ptlrpcd stuck in lnet allocations under memory pressure
Details    : Set PF_MEMALLOC on outgoing path to prevent deadlock on memory
             allocation

Severity   : normal
Bugzilla   : 23575
Description: fix o2iblnd v2 regression of credit deadlock with v1 peers
             (bug 14425).

Severity   : normal
Bugzilla   : 21456
Description: lctl ping of 1.6.6 server from 1.8 router causes protocol error

-------------------------------------------------------------------------------

2010-10-29 Oracle, Inc.
	* version 1.8.5
	* Support for networks:
	  socklnd   - any kernel supported by Lustre,
	  qswlnd    - Qsnet kernel modules 5.20 and later,
	  openiblnd - IbGold 1.8.2,
	  o2iblnd   - OFED 1.3, 1.4.1, 1.4.2, 1.5.1 and 1.5.2
	  viblnd    - Voltaire ibhost 3.4.5 and later,
	  ciblnd    - Topspin 3.2.0,
	  iiblnd    - Infiniserv 3.3 + PathBits patch,
	  gmlnd     - GM 2.1.22 and later,
	  mxlnd     - MX 1.2.10 or later,
	  ptllnd    - Portals 3.3 / UNICOS/lc 1.5.x, 2.0.x

Severity   : normal
Bugzilla   : 23179
Description: MDS node unresponsive
Details    : improve hash distribution, doubled hash size for the lnet cookie
	     hash.

-------------------------------------------------------------------------------

2010-07-31 Oracle, Inc.
	* version 1.8.4
	* Support for networks:
	  socklnd   - any kernel supported by Lustre,
	  qswlnd    - Qsnet kernel modules 5.20 and later,
	  openiblnd - IbGold 1.8.2,
	  o2iblnd   - OFED 1.3, 1.4.1, 1.4.2 and 1.5.1
	  viblnd    - Voltaire ibhost 3.4.5 and later,
	  ciblnd    - Topspin 3.2.0,
	  iiblnd    - Infiniserv 3.3 + PathBits patch,
	  gmlnd     - GM 2.1.22 and later,
	  mxlnd     - MX 1.2.10 or later,
	  ptllnd    - Portals 3.3 / UNICOS/lc 1.5.x, 2.0.x

Severity   : normal
Bugzilla   : 23076
Description: With peer health detection, o2iblnd makes only one attempt to
	     reconnect which is not enough with nodes running lustre 1.6
	     because of proto version mismatch. Fix o2iblnd to retry one more
	     time.


lustre/utils/lctl.c
/*
 * GPL HEADER START
 *
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 only,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License version 2 for more details (a copy is included
 * in the LICENSE file that accompanied this code).
 *
 * You should have received a copy of the GNU General Public License
 * version 2 along with this program; If not, see
 * http://www.sun.com/software/products/lustre/docs/GPLv2.pdf
 *
 * Please contact Sun Microsystems, Inc., 4150 Network Circle, Santa Clara,
 * CA 95054 USA or visit www.sun.com if you need additional information or
 * have any questions.
 *
 * GPL HEADER END
 */
/*
 * Copyright (c) 2002, 2010, Oracle and/or its affiliates. All rights reserved.
 * Use is subject to license terms.
 *
 * Copyright (c) 2012, 2013, Intel Corporation.
 */
/*
 * This file is part of Lustre, http://www.lustre.org/
 * Lustre is a trademark of Sun Microsystems, Inc.
 *
 * lustre/utils/lctl.c
 *
 * Author: Peter J. Braam <braam@clusterfs.com>
 * Author: Phil Schwan <phil@clusterfs.com>
 * Author: Robert Read <rread@clusterfs.com>
 */

#include <stdlib.h>
#include <stdio.h>
#include <lnet/lnetctl.h>
#include "obdctl.h"
#include <libcfs/libcfsutil.h>
#include <lustre/lustre_idl.h>

static int jt_noop(int argc, char **argv) {
        return 0;
}

static int jt_opt_ignore_errors(int argc, char **argv) {
        Parser_ignore_errors(1);
        return 0;
}

command_t cmdlist[] = {
        /* Metacommands */
        {"===== metacommands =======", jt_noop, 0, "metacommands"},
        {"--device", jt_opt_device, 0,
         "run <command> after connecting to device <devno>\n"
         "--device <devno> <command [args ...]>"},
        {"--ignore_errors", jt_opt_ignore_errors, 0,
         "ignore errors that occur during script processing\n"
         "--ignore_errors"},
        {"ignore_errors", jt_opt_ignore_errors, 0,
         "ignore errors that occur during script processing\n"
         "ignore_errors"},

        /* User interface commands */
        {"======== control =========", jt_noop, 0, "control commands"},
        {"help", Parser_help, 0, "help"},
        {"lustre_build_version", jt_get_version, 0,
         "print the build version of lustre\n"
         "usage: lustre_build_version"},
	{"exit", Parser_quit, 0, "quit"},
	{"quit", Parser_quit, 0, "quit"},
	{"--version", Parser_version, 0,
	 "output build version of the utility and exit"},

	/* Network configuration commands */
	{"===== network config =====", jt_noop, 0, "network config"},
        {"--net", jt_opt_net, 0,"run <command> after setting network to <net>\n"
         "usage: --net <tcp/elan/o2ib/...> <command>"},
        {"network", jt_ptl_network, 0, "configure LNET"
         "usage: network up|down"},
        {"net", jt_ptl_network, 0, "configure LNET"
         "usage: net up|down"},
        {"list_nids", jt_ptl_list_nids, 0, "list local NIDs"
         "usage: list_nids [all]"},
        {"which_nid", jt_ptl_which_nid, 0, "choose a NID"
         "usage: which_nid NID [NID...]"},
	{"replace_nids", jt_replace_nids, 0,
	 "replace NIDs for a device\n"
	 "usage: replace_nids <device> <nid1>[,nid2,nid3:nid4,nid5:nid6]"},
	{"clear_conf", jt_clear_configs, 0,
	 "clear config logs for a device or filesystem\n"
	 "usage: clear_conf <device|fsname>"},
        {"interface_list", jt_ptl_print_interfaces,0,"print interface entries\n"
         "usage: interface_list"},
        {"peer_list", jt_ptl_print_peers, 0, "print peer entries\n"
         "usage: peer_list"},
        {"conn_list", jt_ptl_print_connections, 0,
         "print all the connected remote nid\n"
         "usage: conn_list"},
        {"active_tx", jt_ptl_print_active_txs, 0, "print active transmits\n"
         "usage: active_tx"},
        {"route_list", jt_ptl_print_routes, 0,
         "print the portals routing table, same as show_route\n"
         "usage: route_list"},
        {"show_route", jt_ptl_print_routes, 0,
         "print the portals routing table, same as route_list\n"
         "usage: show_route"},
        {"ping", jt_ptl_ping, 0, "Check LNET connectivity\n"
         "usage: ping nid [timeout] [pid]"},

        /* Device selection commands */
        {"==== obd device selection ====", jt_noop, 0, "device selection"},
        {"device", jt_obd_device, 0,
         "set current device to <name|devno>\n"
         "usage: device <%name|$name|devno>"},
        {"device_list", jt_obd_list, 0, "show all devices\n"
         "usage: device_list"},
        {"dl", jt_obd_list, 0, "show all devices\n"
         "usage: dl [-t]"},

        /* Device operations */
        {"==== obd device operations ====", jt_noop, 0, "device operations"},
        {"activate", jt_obd_activate, 0, "activate an import\n"},
        {"deactivate", jt_obd_deactivate, 0, "deactivate an import. "
         "This command should be used on failed OSC devices in an MDT LOV.\n"},
        {"abort_recovery", jt_obd_abort_recovery, 0,
         "abort recovery on a restarting MDT or OST device\n"},
        {"set_timeout", jt_lcfg_set_timeout, 0,
         "usage: conf_param obd_timeout=<secs>\n"},
#if LUSTRE_VERSION_CODE < OBD_OCD_VERSION(3, 0, 53, 0)
        {"conf_param", jt_lcfg_mgsparam, 0,"set a permanent config parameter.\n"
         "This command must be run on the MGS node\n"
         "usage: conf_param [-d] <target.keyword=val>\n"
         "  -d  Remove the permanent setting."},
#endif
        {"local_param", jt_lcfg_param, 0, "set a temporary, local param\n"
         "usage: local_param <target.keyword=val>\n"},
        {"get_param", jt_lcfg_getparam, 0, "get the Lustre or LNET parameter\n"
         "usage: get_param [-n|-N|-F] <param_path1 param_path2 ...>\n"
         "Get the value of Lustre or LNET parameter from the specified path.\n"
         "The path can contain shell-style filename patterns.\n"
         "  -n  Print only the value and not parameter name.\n"
         "  -N  Print only matched parameter names and not the values.\n"
         "      (Especially useful when using patterns.)\n"
         "  -F  When -N specified, add '/', '@' or '=' for directories,\n"
         "      symlinks and writeable files, respectively."},
        {"set_param", jt_lcfg_setparam, 0, "set the Lustre or LNET parameter\n"
	 "usage: set_param [-n] [-P] [-d]"
	 "<param_path1=value1 param_path2=value2 ...>\n"
	 "Set the value of the Lustre or LNET parameter at the specified path.\n"
	 "  -n  Disable printing of the key name when printing values.\n"
	 "  -P  Set the parameter permanently, filesystem-wide.\n"
	 "  -d  Remove the permanent setting (only with -P option).\n"},
	{"list_param", jt_lcfg_listparam, 0,
         "list the Lustre or LNET parameter name\n"
         "usage: list_param [-F|-R] <param_path1 param_path2 ...>\n"
         "List the name of Lustre or LNET parameter from the specified path.\n"
         "  -F  Add '/', '@' or '=' for dirs, symlinks and writeable files,\n"
                "respectively.\n"
         "  -R  Recursively list all parameters under the specified path.\n"},

        /* Debug commands */
        {"==== debugging control ====", jt_noop, 0, "debug"},
        {"debug_daemon", jt_dbg_debug_daemon, 0,
         "debug daemon control and dump to a file\n"
         "usage: debug_daemon {start file [#MB]|stop}"},
        {"debug_kernel", jt_dbg_debug_kernel, 0,
         "get debug buffer and dump to a file, same as dk\n"
         "usage: debug_kernel [file] [raw]"},
        {"dk", jt_dbg_debug_kernel, 0,
         "get debug buffer and dump to a file, same as debug_kernel\n"
         "usage: dk [file] [raw]"},
        {"debug_file", jt_dbg_debug_file, 0,
         "convert a binary debug file dumped by the kernel to ASCII text\n"
         "usage: debug_file <input> [output]"},
        {"df", jt_dbg_debug_file, 0,
         "read debug buffer from input and dump to output, same as debug_file\n"
         "usage: df <input> [output]"},
        {"clear", jt_dbg_clear_debug_buf, 0, "clear kernel debug buffer\n"
         "usage: clear"},
        {"mark", jt_dbg_mark_debug_buf, 0,
         "insert marker text in kernel debug buffer\n"
         "usage: mark <text>"},
        {"filter", jt_dbg_filter, 0, "filter message type\n"
         "usage: filter <subsystem id/debug mask>"},
        {"show", jt_dbg_show, 0, "Show specific type of messages\n"
         "usage: show <subsystem id/debug mask>"},
        {"debug_list", jt_dbg_list, 0, "list subsystem and debug types\n"
         "usage: debug_list <subs/types>"},
        {"modules", jt_dbg_modules, 0,
         "provide gdb-friendly module information\n"
         "usage: modules <path>"},

        /* virtual block operations */
        {"==== virtual block device ====", jt_noop, 0, "virtual block device"},
        {"blockdev_attach", jt_blockdev_attach, 0,
         "attach a lustre regular file to a virtual block device\n"
         "usage: blockdev_attach <file_name> <device_name>"},
        {"blockdev_detach", jt_blockdev_detach, 0,
         "detach a lustre regular file from a virtual block device\n"
         "usage: blockdev_detach <device_name>"},
        {"blockdev_info", jt_blockdev_info, 0,
         "get the device info of a attached file\n"
         "usage: blockdev_info <device_name>"},

        /* Pool commands */
        {"===  Pools ==", jt_noop, 0, "pool management"},
        {"pool_new", jt_pool_cmd, 0,
         "add a new pool\n"
         "usage pool_new <fsname>.<poolname>"},
        {"pool_add", jt_pool_cmd, 0,
         "add the named OSTs to the pool\n"
         "usage pool_add <fsname>.<poolname> <ostname indexed list>"},
        {"pool_remove", jt_pool_cmd, 0,
         "remove the named OST from the pool\n"
         "usage pool_remove <fsname>.<poolname> <ostname indexed list>"},
        {"pool_destroy", jt_pool_cmd, 0,
         "destroy a pool\n"
         "usage pool_destroy <fsname>.<poolname>"},
        {"pool_list", jt_pool_cmd, 0,
         "list pools and pools members\n"
         "usage pool_list  <fsname>[.<poolname>] | <pathname>"},

        /* Changelog commands */
        {"===  Changelogs ==", jt_noop, 0, "changelog user management"},
        {"changelog_register", jt_changelog_register, 0,
         "register a new persistent changelog user, returns id\n"
         "usage:\tdevice <mdtname>\n\tchangelog_register [-n]"},
        {"changelog_deregister", jt_changelog_deregister, 0,
         "deregister an existing changelog user\n"
         "usage:\tdevice <mdtname>\n\tchangelog_deregister <id>"},

        /* Device configuration commands */
        {"== device setup (these are not normally used post 1.4) ==",
                jt_noop, 0, "device config"},
        {"attach", jt_lcfg_attach, 0,
         "set the type, name, and uuid of the current device\n"
         "usage: attach type name uuid"},
        {"detach", jt_obd_detach, 0,
         "remove driver (and name and uuid) from current device\n"
         "usage: detach"},
        {"setup", jt_lcfg_setup, 0,
         "type specific device configuration information\n"
         "usage: setup <args...>"},
        {"cleanup", jt_obd_cleanup, 0, "cleanup previously setup device\n"
         "usage: cleanup [force | failover]"},
        {"dump_cfg", jt_cfg_dump_log, 0,
         "print log of recorded commands for this config to kernel debug log\n"
         "usage: dump_cfg config-uuid-name"},

        /* Test only commands */
        {"==== testing (DANGEROUS) ====", jt_noop, 0, "testing (DANGEROUS)"},
        {"--threads", jt_opt_threads, 0,
         "run <threads> separate instances of <command> on device <devno>\n"
         "--threads <threads> <verbose> <devno> <command [args ...]>"},
        {"lookup", jt_obd_mdc_lookup, 0, "report file mode info\n"
         "usage: lookup <directory> <file>"},
        {"readonly", jt_obd_set_readonly, 0,
         "disable writes to the underlying device\n"},
        {"notransno", jt_obd_no_transno, 0,
         "disable sending of committed-transno updates\n"},
        {"add_uuid", jt_lcfg_add_uuid, 0, "associate a UUID with a nid\n"
         "usage: add_uuid <uuid> <nid>"},
        {"del_uuid", jt_lcfg_del_uuid, 0, "delete a UUID association\n"
         "usage: del_uuid <uuid>"},
        {"add_peer", jt_ptl_add_peer, 0, "add an peer entry\n"
         "usage: add_peer <nid> <host> <port>"},
        {"del_peer", jt_ptl_del_peer, 0, "remove an peer entry\n"
         "usage: del_peer [<nid>] [<ipaddr|pid>]"},
        {"add_conn ", jt_lcfg_add_conn, 0,
         "usage: add_conn <conn_uuid> [priority]\n"},
        {"del_conn ", jt_lcfg_del_conn, 0,
         "usage: del_conn <conn_uuid> \n"},
        {"disconnect", jt_ptl_disconnect, 0, "disconnect from a remote nid\n"
         "usage: disconnect [<nid>]"},
        {"push", jt_ptl_push_connection, 0, "flush connection to a remote nid\n"
         "usage: push [<nid>]"},
        {"mynid", jt_ptl_mynid, 0, "inform the socknal of the local nid. "
         "The nid defaults to hostname for tcp networks and is automatically "
         "setup for elan/myrinet networks.\n"
         "usage: mynid [<nid>]"},
        {"fail", jt_ptl_fail_nid, 0, "fail/restore communications.\n"
         "Omitting the count means indefinitely, 0 means restore, "
         "otherwise fail 'count' messages.\n"
         "usage: fail nid|_all_ [count]"},

        /*Test commands for echo client*/
        {"test_create", jt_obd_test_create, 0,
         "create files on MDT by echo client\n"
         "usage: test_create [-d parent_basedir] <-D parent_count> "
         "[-b child_base_id] <-c stripe_count> <-n count> <-t time>\n"},
        {"test_mkdir", jt_obd_test_mkdir, 0,
         "mkdir on MDT by echo client\n"
         "usage: test_mkdir [-d parent_basedir] <-D parent_count>"
         "[-b child_base_id] [-n count] <-t time>\n"},
        {"test_destroy", jt_obd_test_destroy, 0,
         "Destroy files on MDT by echo client\n"
         "usage: test_destroy [-d parent_basedir] <-D parent_count>"
         "[-b child_base_id] [-n count] <-t time>\n"},
        {"test_rmdir", jt_obd_test_rmdir, 0,
         "rmdir on MDT by echo client\n"
         "usage: test_rmdir [-d parent_basedir] <-D parent_count>"
         "[-b child_base_id] [-n count] <-t time>\n"},
        {"test_lookup", jt_obd_test_lookup, 0,
         "lookup files on MDT by echo client\n"
         "usage: test_lookup [-d parent_basedir] <-D parent_count>"
         "[-b child_base_id] [-n count] <-t time>\n"},
        {"test_setxattr", jt_obd_test_setxattr, 0,
         "Set EA for files/directory on MDT by echo client\n"
         "usage: test_setxattr [-d parent_baseid] <-D parent_count>"
         "[-b child_base_id] [-n count] <-t time>\n"},
        {"test_md_getattr", jt_obd_test_md_getattr, 0,
         "getattr files on MDT by echo client\n"
         "usage: test_md_getattr [-d parent_basedir] <-D parent_count>"
         "[-b child_base_id] [-n count] <-t time>\n"},
        {"getattr", jt_obd_getattr, 0,
         "get attribute for OST object <objid>\n"
         "usage: getattr <objid>"},
        {"setattr", jt_obd_setattr, 0,
         "set mode attribute for OST object <objid>\n"
         "usage: setattr <objid> <mode>"},
        {"create", jt_obd_create, 0,
         "create <num> OST objects (with <mode>)\n"
         "usage: create [num [mode [verbose [lsm data]]]]"},
        {"destroy", jt_obd_destroy, 0,
         "destroy OST object <objid> [num [verbose]]\n"
         "usage: destroy <num> objects, starting at objid <objid>"},
        {"test_getattr", jt_obd_test_getattr, 0,
         "do <num> getattrs (on OST object <objid> (objid+1 on each thread))\n"
         "usage: test_getattr <num> [verbose [[t]objid]]"},
        {"test_setattr", jt_obd_test_setattr, 0,
         "do <num> setattrs (on OST object <objid> (objid+1 on each thread))\n"
         "usage: test_setattr <num> [verbose [[t]objid]]"},
        {"test_brw", jt_obd_test_brw, 0,
         "do <num> bulk read/writes (<npages> per I/O, on OST object <objid>)\n"
         "usage: test_brw [t]<num> [write [verbose [npages [[t]objid]]]]"},
        {"memhog", jt_ptl_memhog, 0,
         "memory pressure testing\n"
         "usage: memhog <page count> [<gfp flags>]"},
        {"getobjversion", jt_get_obj_version, 0,
         "get the version of an object on servers\n"
	 "usage: getobjversion <fid>\n"
	 "	 getobjversion -i <id> -g <group>"},

	/* LFSCK commands */
	{"==== LFSCK ====", jt_noop, 0, "LFSCK"},
	{"lfsck_start", jt_lfsck_start, 0, "start LFSCK\n"
	 "usage: lfsck_start <-M | --device [MDT,OST]_device>\n"
	 "		     [-e | --error error_handle] [-h | --help]\n"
	 "		     [-n | --dryrun switch] [-r | --reset]\n"
	 "		     [-s | --speed speed_limit]\n"
	 "		     [-t | --type lfsck_type[,lfsck_type...]]"},
	{"lfsck_stop", jt_lfsck_stop, 0, "stop lfsck(s)\n"
	 "usage: lfsck_stop <-M | --device [MDT,OST]_device> [-h | --help]"},

	{"==== obsolete (DANGEROUS) ====", jt_noop, 0, "obsolete (DANGEROUS)"},
	/* some test scripts still use these */
	{"cfg_device", jt_obd_device, 0,
	 "set current device to <name>\n"
	 "usage: device <name>"},
	{"recover", jt_obd_recover, 0,
	 "try to restore a lost connection immediately\n"
	 "usage: recover [MDC/OSC device]"},
	/* saving for sanity 44a */
	{"lov_getconfig", jt_obd_lov_getconfig, 0,
	 "read lov configuration from an mds device\n"
	 "usage: lov_getconfig <mountpoint>"},
	/* Llog operations */
	{"llog_catlist", jt_llog_catlist, 0,
	 "list all catalog logs on current device.\n"
	 "usage: llog_catlist"},
	{"llog_info", jt_llog_info, 0,
	 "print log header information.\n"
	 "usage: llog_info <$logname|#oid#ogr#ogen>\n"
	 "       oid, ogr and ogen are hexadecimal."},
	{"llog_print", jt_llog_print, 0,
	 "print log content information.\n"
	 "usage: llog_print <$logname|#oid#ogr#ogen> [from] [to]\n"
	 "       oid, ogr and ogen are hexadecimal.\n"
	 "       print all records from index 1 by default."},
	{"llog_check", jt_llog_check, 0,
	 "print log content information.\n"
	 "usage: llog_check <$logname|#oid#ogr#ogen> [from] [to]\n"
	 "       oid, ogr and ogen are hexadecimal.\n"
	 "       check all records from index 1 by default."},
	 {"llog_cancel", jt_llog_cancel, 0,
	 "cancel one record in log.\n"
	 "This command supports both positional and optional arguments\n"
	 "usage (positional args): "
	 "llog_cancel <catalog id|catalog name> [log id] <index>\n"
	 "usage (optional args): "
	 "llog_cancel --catalog <catalog id|catalog name> --log_id <log_id> "
	 "--log_idx <index>"},
	{"llog_remove", jt_llog_remove, 0,
	 "remove one log from catalog or plain log, erase it from disk.\n"
	 "usage: llog_remove <catalog id|catalog name> <log id>"},
	/* network operations */
	{"add_interface", jt_ptl_add_interface, 0, "add interface entry\n"
	 "usage: add_interface ip [netmask]"},
	{"del_interface", jt_ptl_del_interface, 0, "del interface entry\n"
	 "usage: del_interface [ip]"},
	{"add_route", jt_ptl_add_route, 0,
	 "add an entry to the portals routing table\n"
	 "usage: add_route <gateway> [<hops> [<priority>]]"},
	{"del_route", jt_ptl_del_route, 0,
	 "delete route via gateway to targets from the portals routing table\n"
	 "usage: del_route <gateway> [<target>] [<target>]"},
	{"set_route", jt_ptl_notify_router, 0,
	 "enable/disable routes via gateway in the portals routing table\n"
	 "usage: set_route <gateway> <up/down> [<time>]"},

	{ 0, 0, 0, NULL }
};

int lctl_main(int argc, char **argv)
{
        int rc;

        setlinebuf(stdout);

        ptl_initialize(argc, argv);
        if (obd_initialize(argc, argv) < 0)
                exit(2);
        if (dbg_initialize(argc, argv) < 0)
                exit(3);

	Parser_init("lctl > ", cmdlist);

        if (argc > 1) {
                rc = Parser_execarg(argc - 1, argv + 1, cmdlist);
        } else {
                rc = Parser_commands();
        }

        obd_finalize(argc, argv);
        return rc < 0 ? -rc : rc;
}

#ifndef LIBLUSTRE_TEST
int main (int argc, char **argv)
{
        return (lctl_main (argc, argv));
}
#endif
